abi/codec.go

```diff
@@ -24,9 +33,34 @@ func newCodec(args argsNewCodec) (*codec, error) {
 		return nil, errors.New("cannot create codec: bad public key length")
 	}
 
-	return &codec{
-		pubKeyLength: args.pubKeyLength,
-	}, nil
+	codec := &codec{
+		codecForBool:     &codecForBool{},
+		codecForSmallInt: &codecForSmallInt{},
+		codecForBigInt:   &codecForBigInt{},
+		codecForAddress: &codecForAddress{
+			pubKeyLength: args.pubKeyLength,
+		},
+		codecForString: &codecForString{},
+		codecForBytes:  &codecForBytes{},
+	}
+
+	codec.codecForStruct = &codeForStruct{
+		generalCodec: codec,
+	}
+
+	codec.codecForEnum = &codecForEnum{
+		generalCodec: codec,
+	}
+
+	codec.codecForOption = &codecForOption{
+		generalCodec: codec,
+	}
+
+	codec.codecForList = &codecForList{
+		generalCodec: codec,
```

Comment:
well, this did not turn out as expected. This is actually worse than the initial implementation, Why? We now have reciprocal pointer links which should always be avoided.
Can we move the into a new baseCoded the general functionality from the codec?
Or at least, have an unexported interface instead of clear pointer.

Reply:
Applied the second suggestion (depend on the general codec by means of an interface).

There isn't a "base functionality" per-se that can be extracted to a base codec. That is, some "leaf" codecs (e.g. the one for lists of items) depends on all the other codecs at runtime (to be able to handle all types of items within the list).

Another approach (with a bigger refactoring, though) would be to attach the encoding and decoding functionality to the `value` structs themselves. That way, we would be able to apply the composite pattern as follows (example):

```
struct ListValue:
    encodeNested()
        for each item in the list, call item.encodeNested()
    ...
    
struct U64Value:
    encodeNested():
        ...

list = new ListValue([new U64Value(42), new U64Value(43)])
list.encodeNested(...)
```

However, in this way, we would couple the dummy wrapper values with encoding / decoding logic. E.g. if we decide upon a different serialization format (though chances are slim), a separate family of structs will be required e.g. `ListValueV3`. This would also be possible to overcome using factories - of families of structs (with a given encoding / decoding specification), but that would somehow complicate a bit the design.

What is your opinion? All right as now, using an internal interface for `generalCodec` - or should we switch to another design (e.g. the one in the paragraph above)?

Reply:
Re-re-thinking about it - perhaps having moved the encoding / decoding logic in the `value` structs themselves, and have them implement the `Encodable` interface (or something like that) - with 4 functions (encode nested, encode top level, decode nested, decode top level) - does not sound that bad anymore.

Reply:
That is exactly what we have with the `trie` elements: leafNode, extensionNode, branchNode. The "child"/"children" node(s) is always an interface that can be either implementation

Reply:
Thanks. Added a GH issue here for this topic: https://github.com/multiversx/mx-sdk-abi-go/issues/5